contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,216 @@
+        _;
+    }
+
+    /**
+     * @notice Constructs the Forwarder contract.
+     * @dev _disableInitializers() restricts anybody from initializing the implementation contract, which if not done,
+     * may disrupt the proxy if another EOA were to initialize it.
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initializes the forwarder contract.
+     * @param _crossDomainAdmin L1 address of the contract which can send root bundles/messages to this forwarder contract.
+     * In practice, this is the hub pool for forwarders deployed on L2, and other forwarder contracts on layers != L2.
+     */
+    function __Forwarder_init(address _crossDomainAdmin) public onlyInitializing {
+        __UUPSUpgradeable_init();
+        _setCrossDomainAdmin(_crossDomainAdmin);
+    }
+
+    /**
+     * @notice Sets a new cross domain admin for this contract.
+     * @param _newCrossDomainAdmin L1 address of the new cross domain admin.
+     */
+    function setCrossDomainAdmin(address _newCrossDomainAdmin) external onlyAdmin {
+        _setCrossDomainAdmin(_newCrossDomainAdmin);
+    }
+
+    /**
+     * @notice Sets a new RouteIndex to a specified address.
+     * @param _destinationAddress The address to set in the availableRoutes mapping.
+     * @param _routeIndex RouteIndex struct containing the next immediate path to reach the _destination address.
+     * @dev Each forwarder will not know how many layers it must traverse to arrive at _destinationAddress until it is precisely
+     * one layer away. The availableRoutes mapping lets the forwarder know what path it should take to progess, and then relies on
+     * the forwarders on the subsequent layers to derive the next route in sequence.
+     */
+    function updateRoute(address _destinationAddress, RouteIndex memory _routeIndex) external onlyAdmin {
+        availableRoutes[_destinationAddress] = _routeIndex;
+        emit RouteUpdated(_destinationAddress, _routeIndex);
+    }
+
+    /**
+     * @notice Sets a new remote token in the remoteTokens mapping.
+     * @param _destinationAddress The address to set in the remoteTokens mapping. This is to identify the network associated with the
+     * remote token.
+     * @param _baseToken The address of the token which exists on the current network.
+     * @param _remoteToken The address of the token on the network which is next on the path to _destinationAddress.
+     * @dev This mapping also relies on using _destinationAddress to determine the network to bridge to.
+     */
+    function updateRemoteToken(
+        address _destinationAddress,
+        address _baseToken,
+        address _remoteToken
+    ) external onlyAdmin {
+        remoteTokens[_destinationAddress][_baseToken] = _remoteToken;
+        emit RemoteTokensUpdated(_destinationAddress, _baseToken, _remoteToken);
+    }
+
+    /**
+     * @notice Relays a specified message to a target on an arbitrary layer.
```

Comment:
This is not forwarding to an "arbitrary" layer, though, is it? There is a hardcoded next layer to target messages to and depending on the `target` address, this contract changes its behavior

Reply:
https://github.com/across-protocol/contracts/pull/609/commits/d33099343f18b48c9db005412529ced8980b7c86